dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handling/DataHandler.java

```diff
@@ -0,0 +1,1161 @@
+            .withEarliestStartDateLatestEndDate()
+            .withPeriodDimensionWithoutOptions()
+            .withIncludePeriodStartEndDates( true );
+
+        if ( params.isShowHierarchy() )
+        {
+            builder.withOrgUnitLevels( organisationUnitService.getFilledOrganisationUnitLevels() );
+        }
+
+        return builder.build();
+    }
+
+    /**
+     * Generates a mapping between the count of a validation result.
+     *
+     * @param params the {@link DataQueryParams}.
+     * @return a mapping between validation results and counts of them
+     */
+    private Map<String, Double> getAggregatedValidationResultMapObjectTyped( DataQueryParams params )
+    {
+        return getDoubleMap( getAggregatedValueMap( params, VALIDATION_RESULT, newArrayList() ) );
+    }
+
+    /**
+     * Fill grid with aggregated data map with key and value
+     *
+     * @param params the {@link DataQueryParams}.
+     * @param grid the grid
+     * @param aggregatedDataMap the aggregated data map
+     */
+    private void fillGridWithAggregatedDataMap( DataQueryParams params, Grid grid,
+        Map<String, Double> aggregatedDataMap )
+    {
+        for ( Map.Entry<String, Double> entry : aggregatedDataMap.entrySet() )
+        {
+            Double value = params.isSkipRounding() ? entry.getValue() : getRounded( entry.getValue() );
+
+            grid.addRow()
+                .addValues( entry.getKey().split( DIMENSION_SEP ) )
+                .addValue( value );
+
+            if ( params.isIncludeNumDen() )
+            {
+                grid.addNullValues( NUMERATOR_DENOMINATOR_PROPERTIES_COUNT );
+            }
+        }
+    }
+
+    /**
+     * Adds reporting rates to the given grid based on the given data query
+     * parameters and reporting rate metric.
+     *
+     * @param params the {@link DataQueryParams}.
+     * @param grid the grid.
+     * @param metric the reporting rate metric.
+     */
+    private void addReportingRates( DataQueryParams params, Grid grid, ReportingRateMetric metric )
+    {
+        if ( !params.getReportingRates().isEmpty() && !params.isSkipData() )
+        {
+            if ( !COMPLETENESS_DIMENSION_TYPES.containsAll( params.getDimensionTypes() ) )
+            {
+                return;
+            }
+
+            DataQueryParams targetParams = newBuilder( params )
+                .withSkipPartitioning( true )
+                .withTimely( false )
+                .withRestrictByOrgUnitOpeningClosedDate( true )
+                .withRestrictByCategoryOptionStartEndDate( true )
+                .withAggregationType( SUM ).build();
+
+            Map<String, Double> targetMap = getAggregatedCompletenessTargetMap( targetParams );
+
+            Map<String, Double> dataMap = metric != EXPECTED_REPORTS
+                ? getAggregatedCompletenessValueMap( params )
+                : new HashMap<>();
+
+            Integer periodIndex = params.getPeriodDimensionIndex();
```

Comment:
can we use a primitive (`int`) here?

Reply:
I'm going to have a look. I didn't want to change any code to avoid any risk of introducing bugs. Let me check if this is safe enough.

Reply:
I didn't feel confident about this change, as we could have some null checking that would not be the same as 0 (zero). Not sure if this would possible, but I'm afraid it could introduce some bug. I think it's better to leave for another round of refactoring

Reply:
ok